kroxylicious-api/src/main/java/io/kroxylicious/proxy/service/Contributor.java

```diff
@@ -5,48 +5,47 @@
  */
 package io.kroxylicious.proxy.service;
 
-import io.kroxylicious.proxy.config.BaseConfig;
+import edu.umd.cs.findbugs.annotations.NonNull;
 
 /**
- * Support loading an Instance of a service, optionally providing it with configuration obtained
- * from the Kroxylicious configuration file.
+ * Support creating an instance of a service, optionally providing it with configuration.
  *
- * @param <T> the service type
+ * @param <S> the service type
+ * @param <C> the type of config provided to the service, or {@link Void} for config-less service implementations.
+ * @param <X> the context type
  */
-public interface Contributor<T, S extends Context> {
+public interface Contributor<S, C, X extends Context<C>> {
```

Comment:
Do we need `X` here? We could just declare `createInstance(Context<C>)`.

TBH I'm unconvinced that we really need this `Contributor` top level interface. It makes all the plugins look the same, which has pros and cons. It does mean we have a single `ContributionManager` which is a minor convenience for us as implementer of the proxy. But it has some significant cons:
* It forces us into using generics. But we don't even get the full benefit of them because `ContributionManager` is using raw `Contributor`. 
* The existence of `Contributor` is apparent to filter authors, even though it's really an implementation detail. 
* We can't really specify a complete contract for `Contributor` (it's too abstract). For example, although Contributor lets us instantiate it cannot specify any contract to its clients about any responsibility they might have for invoking a `close()` method.  
* We need an abstract `Context` (again, for implementation reasons only) but a filter author might legitimately wonder why all contexts should inherit from the same base type. 

By having these very abstract types like `Contributor` and `Context` we're forced to document them in a way that's meaningful to filter authors (who have to then try to understand these things). TBH I have a hard time explaining what they are or why they're needed in terms that would make sense.

Reply:
Given our use-cases for Contributor are always contributor specific (e.g. FilterConfigTypeIdResolver is only interested in FilterContributors, MicrometerDefinitionBuilder is only interested in MicrometerConfigurationHookContributor), I'm +1 to doing away with the ContributionManager/Contributor/Context.   We'd need something contributor specific to replace ContributionManager e.g. a FilterContributionManager.  If we can't stomach the repetition in n  ContributionManagers, we could consider codegening them with the help of some compile time annotations on the type specific Contributor interfaces.

Reply:
Sam, Rob and I had an out-of-band discussion about this which I should mention here. A couple of ideas...

We might be able to have some common `ContributorManager` code, it would just might work in terms of method references (e.g. to `createInstance`) rather than `Contributor`.  

By having FilterContributor as the only abstraction means we might be able to support a more sophisticated pattern for filter creation. Right now the filter just creates instances. We've not documented the contract for FilterContributor to say whether a single contributor instance creates all filter instances, or whether it only creates one. I don't think we need to document that yet, but when we did it would allow us to add a method wh  

```java 
interface FilterContributor<C, X, F extends Filter> {
  /** The type of configuration this contributor consumes */
  Class<C> configType();

  /** The type of filter that this contributor produces */
  Class<F> filterType();

  /** Called by the runtime once per filter contributor instance to create a common context.
    * The returned context will be passed to {@link #createFilter(C, X)}.
    * {@link #destroyContext(X)} will be called with the returned context when the contributor is shutdown. 
    */
  X createContext(C config);

  /** Called by the runtime (multiple times per filter contributor instance) to create a new filter instance.
   * The context is the one returned by {@link #createContext(C).
   * When the returned filter is no longer needed {@link #destroyFilter(f)} will be called. 
   */
  F createFilter(C config, X context);

  /**
   * Called by the runtime to destroy a filter originally returned by {@link #createFilter(F).
   */
  void destroyFilter(F filter);
 
  /**
   * Called by the runtime to destroy the context originally returned by {@link #createContext(F).
   */
  void destroyContext(X context);
}
```

Obviously this is a bit of a sketch and we can argue over the details:

* The context could be used to, for example, create a connection pool to some external system that is then shared by all the filter instances. 
* We _could_ allow these contexts to support `Closeable`, but not all contexts would need closing, and I don't like the `instanceof Closeable` that would be required. Instead the pattern I'm using here is for the contributor to supply a method to do the closing (which could be defaulted to a no-op).  I don't think it really makes much difference to the filter author. 
* A similar pattern is being used for filters themselves. The contract doesn't specify when we might destroy a Filter, just that the author can rely on filters being destroyed when they're no longer needed.  
* I'm glossing over any additional context that the runtime might want to pass to `createFilter`. Currently we have the `FilterConstructionContext` to which we can add methods in the future. We could support a similar thing here. But I think we make our lives simpler by having separate parameters for the contributors context and the runtimes context. It means the runtime's context doesn't need to be generic, for a start. 
* Finally, I don't think we should implement this _right now_. I'm just putting it forward as a possible way we could evolve this interface in the future. The extra methods would be `default` no-ops, so such evolution would be backwards compatible.

Reply:
One other salient detail to highlight from our out of band discussion and implicit in the above proposal is the life cycle of contributors. 

We need to ensure that contributors are loaded once and not on demand, they are factories of filters and thus have longer lifecycles than connections. Thus `createContext` and `destroyContext` should be called during run time startup and shutdown rather than on connection open or close (as I think the current contributor lifecycle does).

Reply:
Would be keen to defer adding lifecycle methods and the per-Factory Context object to another PR/discussion, so far this PR is a refactor with no new capabilities added.

Reply:
Yes, I agree we should seek to merge this and have further discussion on this idea.